### PR TITLE
Add Golang Queries

### LINKED ISSUE #4: queries
- [x] isMostRecentAncestor
- [x] getRelayGenesis
- [x] getLastReorgCommonAncestor
- [x] ~~findHeight~~ not needed b/c header objects store authenticated height now
- [x] findAncestor
- [x] isAncestor
- [x] heaviestFromAncestor

### PR DESCRIPTION
**Pull request by ErinHales**
Closes #4

### CODE REVIEW

File: golang/x/relay/keeper/chain.go [hunk 1]
```diff
@@ -70,13 +70,13 @@ func (k Keeper) GetBestKnownDigest(ctx sdk.Context) (types.Hash256Digest, sdk.Er
 	return k.getDigestByStoreKey(ctx, types.BestKnownDigestStorage)
 }
 
-// setLastReorgLCA sets the latest common ancestor of the last reorg
```
Comment (prestwich) [inline golang/x/relay/keeper/chain.go]: don't rename this :+1:

File: golang/go.mod [hunk 2]
```diff
@@ -4,6 +4,7 @@ go 1.12
 
 require (
 	github.com/cosmos/cosmos-sdk v0.37.0
+	github.com/go-logfmt/logfmt v0.4.0 // indirect
```
Comment (prestwich) [inline golang/go.mod]: I wonder why yours keeps adding this back :thinking:

Comment (prestwich) [reply golang/go.mod]: mine keeps removing it

File: golang/x/relay/keeper/querier.go [hunk 3]
```diff
@@ -34,8 +60,14 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
 
 	// getAncestor is removed by the handler switch/case block above
 
+	// check that the path is this many items long, error if not
+	if len(path) > 3 {
+		return []byte{}, types.ErrTooManyArguments(types.DefaultCodespace)
+	} else if len(path) < 3 {
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: if the path is 2 items long, accept it, but don't call `strconv.Atoi(path[2])` later. instead, use `15` as a default

File: golang/x/relay/keeper/querier.go [hunk 4]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
 		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
 	}
 
-	// TODO: check that the path is this many items long, error if not
-	limit := uint32(25) // TODO: parse from path, use 240 as default if not in path
+	limit, convErr := strconv.Atoi(path[2]) // TODO: parse from path, use 240 as default if not in path
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: use `ParseUint` instead

File: golang/x/relay/types/errors.go [hunk 5]
```diff
@@ -104,6 +104,23 @@ const (
...
+
+	// 500-block Queries
+
+	// NotEnoughArguments means there are not enough arguments specified in the path of a query
+	NotEnoughArguments sdk.CodeType = 501
+	// NotEnoughArgumentsMessage is the corresponding message
+	NotEnoughArgumentsMessage = "Not enough arguments"
+
+	// TooManyArguments means there are too many arguments specified in the path of a query
+	TooManyArguments sdk.CodeType = 501
```
Comment (prestwich) [inline golang/x/relay/types/errors.go]: reused code number

File: golang/x/relay/keeper/querier.go [hunk 6]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+	res, marshalErr := codec.MarshalJSONIndent(keeper.cdc, response)
+	if marshalErr != nil {
+		return []byte{}, types.ErrMarshalJSON(types.DefaultCodespace)
+	}
+	return res, nil
+}
+
+func queryGetLastReorgCA(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) (res []byte, err sdk.Error) {
+	// This calls the keeper and gets an answer
+	result, err := keeper.GetRelayGenesis(ctx)
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: wrong call here

File: golang/x/relay/keeper/querier.go [hunk 7]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+	digestBytes, decodeErr := hex.DecodeString(hexDigest)
+	if decodeErr != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	digestLE, newDigestErr := btcspv.NewHash256Digest(digestBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	offset, convErr := strconv.Atoi(path[1])
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: use 'parseUint` instead. lets you cut out the `uint32()` cast later

File: golang/x/relay/keeper/querier.go [hunk 8]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+	if convErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, convErr)
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: this is not a BTCSPV error

File: golang/x/relay/keeper/querier.go [hunk 9]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
 		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
 	}
 
-	// TODO: check that the path is this many items long, error if not
-	limit := uint32(25) // TODO: parse from path, use 240 as default if not in path
+	limit, convErr := strconv.Atoi(path[2]) // TODO: parse from path, use 240 as default if not in path
+	if convErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, convErr)
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: this is not a BTCSPV error

File: golang/x/relay/keeper/querier.go [hunk 10]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+	}
+	ancestor, newDigestErr := btcspv.NewHash256Digest(ancestorBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	// parse the second path item as hex.
+	currentBest := path[1]
+	currentBestBytes, decodeErr := hex.DecodeString(currentBest)
+	if err != nil {
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: wrong error checked here

File: golang/x/relay/keeper/querier.go [hunk 11]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	currentBestDigest, newDigestErr := btcspv.NewHash256Digest(currentBestBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	newBest := path[2]
+	newBestBytes, decodeErr := hex.DecodeString(newBest)
+	if err != nil {
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: wrong error checked here

File: golang/x/relay/keeper/querier.go [hunk 12]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	newBestDigest, newDigestErr := btcspv.NewHash256Digest(newBestBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	limit, convErr := strconv.Atoi(path[3])
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: use `ParseUint` instead

also, if path is only 3 long, use a default number, as above

File: golang/x/relay/keeper/querier.go [hunk 13]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	leftDigest, newDigestErr := btcspv.NewHash256Digest(leftBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	right := path[2]
+	rightBytes, decodeErr := hex.DecodeString(right)
+	if err != nil {
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: wrong error checked

File: golang/x/relay/keeper/querier.go [hunk 14]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+	}
+	ancestor, newDigestErr := btcspv.NewHash256Digest(ancestorBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	// parse the second path item as hex.
+	left := path[1]
+	leftBytes, decodeErr := hex.DecodeString(left)
+	if err != nil {
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: wrong error checked

File: golang/x/relay/keeper/querier.go [hunk 15]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+func queryIsMostRecentCommonAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keeper Keeper) (res []byte, err sdk.Error) {
+	// check that the path is this many items long, error if not
+	if len(path) > 4 {
+		return []byte{}, types.ErrTooManyArguments(types.DefaultCodespace)
+	} else if len(path) < 4 {
+		return []byte{}, types.ErrNotEnoughArguments(types.DefaultCodespace)
+	}
+
+	ancestorHex := path[0]
+	ancestorBytes, decodeErr := hex.DecodeString(ancestorHex)
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: we should maybe write a convenience function for making a `hash256digest` from hex

File: golang/x/relay/keeper/querier.go [hunk 16]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	rightDigest, newDigestErr := btcspv.NewHash256Digest(rightBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	limit, convErr := strconv.Atoi(path[3])
+	if convErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, convErr)
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: this is not a BTCSPV error

File: golang/x/relay/keeper/querier.go [hunk 17]
```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+	rightBytes, decodeErr := hex.DecodeString(right)
+	if err != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	rightDigest, newDigestErr := btcspv.NewHash256Digest(rightBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	limit, convErr := strconv.Atoi(path[3])
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: use `ParseUint` instead

also, if path is only 3 long, use a default number, as above

File: golang/x/relay/types/querier.go [hunk 18]
```diff
@@ -3,18 +3,88 @@ package types
...
 	"encoding/hex"
 	"fmt"
-	"strings"
 )
 
 type QueryResIsAncestor struct {
 	Digest              Hash256Digest `json:"digest"`
 	ProspectiveAncestor Hash256Digest `json:"prospectiveAncestor"`
-	IsAncestor          bool          `json:"isAncestor"`
+	Res                 bool          `json:"res"`
```
Comment (prestwich) [inline golang/x/relay/types/querier.go]: let's call this `json:"result"` instead of `json:"res"`

File: golang/x/relay/keeper/chain.go (hunk 1 above)
Comment (ErinHales) [reply golang/x/relay/keeper/chain.go]: Fixed

File: golang/x/relay/types/errors.go (hunk 5 above)
Comment (ErinHales) [reply golang/x/relay/types/errors.go]: Fixed.  Also realized I was returning the wrong thing in the functions below

File: golang/x/relay/keeper/querier.go (hunk 14 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: Fixed

File: golang/x/relay/keeper/querier.go (hunk 3 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: Fixed

File: golang/x/relay/keeper/querier.go (hunk 4 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: Fixed, changed all instances of this to use ParseUint

File: golang/x/relay/keeper/querier.go (hunk 16 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: Ok, I made another function that converts other external errors to sdk errors

File: golang/x/relay/keeper/querier.go (hunk 15 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: Done

File: golang/go.mod (hunk 2 above)
Comment (ErinHales) [reply golang/go.mod]: No idea...

File: golang/x/relay/keeper/querier.go (hunk 17 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: Does it matter what default number? I think you mentioned 240 before...

Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: But earlier we used 15

File: golang/x/relay/genesis.go [hunk 19]
```diff
@@ -52,10 +52,16 @@ func DefaultGenesisState() GenesisState {
 }
 
 // InitGenesis inits the app state based on the genesis state
-func InitGenesis(ctx sdk.Context, keeper Keeper, data GenesisState) []abci.ValidatorUpdate {
-	keeper.SetGenesisState(ctx, data.Headers[0], data.PeriodStart)
-	keeper.IngestHeaderChain(ctx, data.Headers[1:])
-	return []abci.ValidatorUpdate{}
+func InitGenesis(ctx sdk.Context, keeper Keeper, data GenesisState) ([]abci.ValidatorUpdate, sdk.Error) {
+	err := keeper.SetGenesisState(ctx, data.Headers[0], data.PeriodStart)
+	if err != nil {
```
Comment (prestwich) [inline golang/x/relay/genesis.go]: go ahead and revert this, it's changed in my branch

File: golang/x/relay/keeper/querier.go (hunk 6 above)
Comment (prestwich) [reply golang/x/relay/keeper/querier.go]: > queryGetLastReorgLCA

> GetRelayGenesis

File: golang/x/relay/genesis.go (hunk 19 above)
Comment (ErinHales) [reply golang/x/relay/genesis.go]: Ok, reverted

File: golang/x/relay/keeper/keeper_test.go [hunk 20]
```diff
@@ -86,6 +86,7 @@ func logIfTestCaseError(tc NamedCase, err sdk.Error) {
 // Runs the whole test suite
 func TestKeeper(t *testing.T) {
 	jsonFile, err := os.Open("../../../../testVectors.json")
+	logIfError(err)
 	defer jsonFile.Close()
 	logIfError(err)
```
Comment (prestwich) [inline golang/x/relay/keeper/keeper_test.go]: the second logIfError is superfluous

File: golang/x/relay/keeper/querier.go [hunk 21]
```diff
@@ -34,48 +73,242 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
-	if newDigestErr != nil {
-		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	ancestor, ancestorErr := hash256DigestFromHex(path[1])
+	if ancestorErr != nil {
+		return []byte{}, ancestorErr
 	}
 
-	// TODO: check that the path is this many items long, error if not
-	limit := uint32(25) // TODO: parse from path, use 240 as default if not in path
+	var limit uint32
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: good spot for a convenience function. should take

```
func DecodeUint32FromPath(path []string, idx int, default uint32) (uint32, sdk.Error) {
   if len(path) >= idx {
        return default, nil
   }
   // parse int from path[idx], return error if necessary
}
```

File: golang/x/relay/keeper/querier.go [hunk 22]
```diff
@@ -34,48 +73,242 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
...
+	} else if len(path) < 2 {
+		return []byte{}, types.ErrNotEnoughArguments(types.DefaultCodespace)
+	}
+
+	digestLE, digestErr := hash256DigestFromHex(path[0])
+	if digestErr != nil {
+		return []byte{}, digestErr
+	}
+
+	offset, convErr := strconv.ParseUint(path[1], 0, 10)
```
Comment (prestwich) [inline golang/x/relay/keeper/querier.go]: the 3rd argument of `ParseUint` is the number of bits the number fits in. it should be `32` for all of these

File: golang/x/relay/keeper/querier.go (hunk 6 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: Oops

File: golang/x/relay/types/querier.go [hunk 23]
```diff
@@ -3,18 +3,88 @@ package types
...
 	"encoding/hex"
 	"fmt"
-	"strings"
 )
 
 type QueryResIsAncestor struct {
 	Digest              Hash256Digest `json:"digest"`
 	ProspectiveAncestor Hash256Digest `json:"prospectiveAncestor"`
-	IsAncestor          bool          `json:"isAncestor"`
+	Res                 bool          `json:"result"`
```
Comment (prestwich) [inline golang/x/relay/types/querier.go]: let's add `Limit` to this struct, as the `result` may be a false-negative if the limit is low :thinking:

File: golang/x/relay/keeper/keeper_test.go (hunk 20 above)
Comment (ErinHales) [reply golang/x/relay/keeper/keeper_test.go]: I think this has been fixed already

File: golang/x/relay/keeper/querier.go (hunk 21 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: ok, added

File: golang/x/relay/keeper/querier.go (hunk 22 above)
Comment (ErinHales) [reply golang/x/relay/keeper/querier.go]: Oh, fixed

File: golang/x/relay/types/querier.go (hunk 23 above)
Comment (ErinHales) [reply golang/x/relay/types/querier.go]: Added

**Pull request merged**